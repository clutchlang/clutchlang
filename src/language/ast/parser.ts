import * as tokens from './token';

/**
 * Represents an unresolved node in the abstract syntax tree.
 */
export abstract class AstNode {
  public abstract accept<R, C>(
    visitor: AstVisitor<R, C>,
    context?: C
  ): R | undefined;

  /**
   * First token that represents this AST node.
   */
  public abstract get first(): tokens.Token;

  /**
   * Last token that represents this AST node.
   */
  public abstract get last(): tokens.Token;
}

/**
 * Represents a simple AST node that originates from a single token.
 */
export abstract class SimpleNode extends AstNode {
  constructor(protected readonly token: tokens.Token) {
    super();
  }

  public get first(): tokens.Token {
    return this.token;
  }

  public get last(): tokens.Token {
    return this.token;
  }
}

/**
 * Represents an operator node parsed from a single token.
 */
export class Operator extends SimpleNode {
  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitOperator(this, context);
  }
}

/**
 * Represents a marker type of AST node for all statements.
 */
export abstract class Statement extends AstNode {}

/**
 * Represents a marker type of AST node for all expressions.
 */
export abstract class Expression extends Statement {}

/**
 * Represents a simple AST node that is also an expression.
 */
export abstract class SimpleExpression extends Expression {
  constructor(protected readonly token: tokens.Token) {
    super();
  }

  public get first(): tokens.Token {
    return this.token;
  }

  public get last(): tokens.Token {
    return this.token;
  }
}

/**
 * Represents an expression that utilizes an operator.
 */
export abstract class OperatorExpression extends Expression {
  constructor(
    public readonly operator: Operator,
    public readonly target: Expression
  ) {
    super();
  }
}

/**
 * Represents an expression in the form of `<OP><EXPR>`.
 */
export class PrefixExpression extends OperatorExpression {
  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitPrefixExpression(this, context);
  }

  public get first(): tokens.Token {
    return this.operator.first;
  }

  public get last(): tokens.Token {
    return this.target.last;
  }
}

/**
 * Represents an expression in the form of `<EXPR><OP>`.
 */
export class PostfixExpression extends OperatorExpression {
  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitPostfixExpression(this, context);
  }

  public get first(): tokens.Token {
    return this.target.first;
  }

  public get last(): tokens.Token {
    return this.operator.last;
  }
}

/**
 * Represents an expression in the form of `<EXPR><OP><EXPR>`.
 */
export class BinaryExpression extends OperatorExpression {
  constructor(
    operator: Operator,
    public readonly left: Expression,
    public readonly right: Expression
  ) {
    super(operator, left);
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitBinaryExpression(this, context);
  }

  public get first(): tokens.Token {
    return this.left.first;
  }

  public get last(): tokens.Token {
    return this.right.last;
  }
}

/**
 * Represents an expression in the form of `<TARGET>.<PROPERTY>`.
 */
export class PropertyExpression extends Expression {
  constructor(
    public readonly target: Expression,
    public readonly property: Identifier
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitPropertyExpression(this, context);
  }

  public get first(): tokens.Token {
    return this.target.first;
  }

  public get last(): tokens.Token {
    return this.property.last;
  }
}

/**
 * Represents a list of arguments passed to a function call.
 */
export class ArgumentList extends AstNode {
  constructor(
    public readonly first: tokens.Token,
    public readonly args: Expression[],
    public readonly last: tokens.Token
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitArgumentList(this, context);
  }
}

/**
 * Represents an expression for a function call.
 *
 * May _optionally_ contain a @member target.
 */
export class CallExpression extends Expression {
  constructor(
    public readonly name: Identifier,
    public readonly args: ArgumentList,
    public readonly target?: Expression
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitCallExpression(this, context);
  }

  public get first(): tokens.Token {
    return (this.target || this.name).first;
  }

  public get last(): tokens.Token {
    return this.args.last;
  }
}

export class ConditionalExpression extends Expression {
  constructor(
    public readonly first: tokens.Token,
    public readonly condition: Expression,
    public readonly body: Expression | StatementBlock,
    public readonly elseCondition?: Expression,
    public readonly elseBody?: Expression | StatementBlock
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitConditionalExpression(this, context);
  }

  public get last(): tokens.Token {
    return (this.elseBody || this.body).last;
  }
}

export class GroupExpression extends Expression {
  constructor(
    public readonly first: tokens.Token,
    public readonly expression: Expression,
    public readonly last: tokens.Token
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitGroupExpression(this, context);
  }
}

/**
 * Represents an expression for a literal identifier (name).
 */
export class Identifier extends SimpleExpression {
  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitIdentifier(this, context);
  }

  public get name(): string {
    return this.token.lexeme;
  }
}

/**
 * Represents an expression for a literal boolean.
 */
export class LiteralBoolean extends SimpleExpression {
  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitLiteralBoolean(this, context);
  }

  public get value(): string {
    return this.token.lexeme;
  }
}

/**
 * Represents an expression for a literal number.
 */
export class LiteralNumber extends SimpleExpression {
  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitLiteralNumber(this, context);
  }

  public get value(): string {
    return this.token.lexeme;
  }
}

/**
 * Represents an expression for a literal string.
 */
export class LiteralString extends SimpleExpression {
  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitLiteralString(this, context);
  }

  public get value(): string {
    return this.token.lexeme;
  }
}

/**
 * Represents a variable declared anywhere in scope.
 */
export class VariableDeclaration extends AstNode {
  constructor(
    public readonly name: Identifier,
    public readonly type?: Identifier,
    public readonly value?: Expression
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitVariableDeclaration(this, context);
  }

  public get first(): tokens.Token {
    return this.name.first;
  }

  public get last(): tokens.Token {
    return (this.value || this.type || this.name).last;
  }
}

/**
 * Represents a list of parameters for a function or closure definition.
 */
export class ParameterList extends AstNode {
  constructor(
    public readonly first: tokens.Token,
    public readonly params: VariableDeclaration[],
    public readonly last: tokens.Token
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitParameterList(this, context);
  }
}

/**
 * Represents a series of statements.
 */
export class StatementBlock extends AstNode {
  constructor(
    public readonly first: tokens.Token,
    public readonly statements: Statement[],
    public readonly last: tokens.Token
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitStatementBlock(this, context);
  }
}

/**
 * Represents a return statement.
 */
export class ReturnStatement extends AstNode {
  constructor(
    public readonly first: tokens.Token,
    public readonly expression?: Expression
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitReturnStatement(this, context);
  }

  public get last(): tokens.Token {
    return this.expression ? this.expression.last : this.first;
  }
}

/**
 * Represents a function declared anywhere in scope.
 */
export class FunctionDeclaration extends AstNode {
  constructor(
    public readonly name: Identifier,
    public readonly params: ParameterList,
    public readonly body: Expression | StatementBlock,
    public readonly returnType?: Identifier
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitFunctionDeclaration(this, context);
  }

  public get first(): tokens.Token {
    return this.name.first;
  }

  public get last(): tokens.Token {
    return this.body.last;
  }
}

/**
 * Represents a type declared anywhere in scope.
 */
export class TypeDeclaration extends AstNode {
  constructor(
    public readonly modifiers: tokens.Token[],
    public readonly name: Identifier,
    public readonly members: Array<FunctionDeclaration | VariableDeclaration>,
    public readonly last: tokens.Token
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitTypeDeclaration(this, context);
  }

  public get first(): tokens.Token {
    return this.modifiers.length ? this.modifiers[0] : this.name.first;
  }
}

/**
 *
 */
export class ModuleRoot extends AstNode {
  constructor(
    public readonly declarations: Array<
      FunctionDeclaration | TypeDeclaration | VariableDeclaration
    >
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitModuleRoot(this, context);
  }

  public get first(): tokens.Token {
    return this.declarations[0].first;
  }

  public get last(): tokens.Token {
    return this.declarations[this.declarations.length - 1].last;
  }
}

export class AstParser {}

/**
 * An AST visitor that by default does nothing visiting an AST node.
 *
 * It is intended to be a superclass for classes that use the visitor pattern
 * primarily as a dispatch mechanism (and do not need to recursively need to
 * visit the whole structure) and only need to visit a small number of node
 * types.
 */
export class AstVisitor<R, C> {
  public visitArgumentList(_: ArgumentList, __?: C): R | undefined {
    return;
  }

  public visitBinaryExpression(_: BinaryExpression, __?: C): R | undefined {
    return;
  }

  public visitCallExpression(_: CallExpression, __?: C): R | undefined {
    return;
  }

  public visitConditionalExpression(
    _: ConditionalExpression,
    __?: C
  ): R | undefined {
    return;
  }

  public visitFunctionDeclaration(
    _: FunctionDeclaration,
    __?: C
  ): R | undefined {
    return;
  }

  public visitGroupExpression(_: GroupExpression, __?: C): R | undefined {
    return;
  }

  public visitIdentifier(_: Identifier, __?: C): R | undefined {
    return;
  }

  public visitLiteralBoolean(_: LiteralBoolean, __?: C): R | undefined {
    return;
  }

  public visitLiteralNumber(_: LiteralNumber, __?: C): R | undefined {
    return;
  }

  public visitLiteralString(_: LiteralString, __?: C): R | undefined {
    return;
  }

  public visitModuleRoot(_: ModuleRoot, __?: C): R | undefined {
    return;
  }

  public visitOperator(_: Operator, __?: C): R | undefined {
    return;
  }

  public visitParameterList(_: ParameterList, __?: C): R | undefined {
    return;
  }

  public visitPrefixExpression(_: PrefixExpression, __?: C): R | undefined {
    return;
  }

  public visitPostfixExpression(_: PostfixExpression, __?: C): R | undefined {
    return;
  }

  public visitPropertyExpression(_: PropertyExpression, __?: C): R | undefined {
    return;
  }

  public visitReturnStatement(_: ReturnStatement, __?: C): R | undefined {
    return;
  }

  public visitStatementBlock(_: StatementBlock, __?: C): R | undefined {
    return;
  }

  public visitTypeDeclaration(_: TypeDeclaration, __?: C): R | undefined {
    return;
  }

  public visitVariableDeclaration(
    _: VariableDeclaration,
    __?: C
  ): R | undefined {
    return;
  }
}
