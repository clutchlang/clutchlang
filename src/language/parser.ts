export * from './parser/expressions';
export * from './parser/factory';
export * from './parser/nodes';
export * from './parser/operators';
export * from './parser/statements';
export * from './parser/visitors';
