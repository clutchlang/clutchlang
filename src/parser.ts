export { Codes } from './parser/codes';
export { Lexer } from './parser/lexer';
export { Token } from './parser/token';
export { Tokens } from './parser/tokens';
