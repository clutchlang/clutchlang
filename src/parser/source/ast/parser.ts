import {
  RegExpToken,
  StringToken,
  SymbolToken,
  Token,
  TokenKind,
} from '../tokenizer/tokens';
import {
  AstCompilationUnit,
  AstExpression,
  AstFunctionDeclaration,
  AstInvocationExpression,
  AstLiteralBoolean,
  AstLiteralIdentifier,
  AstLiteralNumber,
  AstLiteralString,
  AstParenthesizedExpression,
} from './node';
import { TokenScanner } from './scanner';

export class AstParser {
  constructor(private readonly scanner: TokenScanner) {}

  public parseCompilationUnit(): AstCompilationUnit {
    const results: AstFunctionDeclaration[] = [];
    while (!this.scanner.isDone) {
      // TODO: Loosen once alternatives are supported.
      this.scanRequired(RegExpToken.Identifier);
      results.push(this.parseFunctionDeclaration());
    }
    return new AstCompilationUnit(results);
  }

  /**
   * Returns a parsed expression, or undefined if there is none.
   */
  private parseExpression(): AstExpression | undefined {
    const token = this.scanner.peek();
    switch (token.kind) {
      case SymbolToken.LParen:
        this.scanRequired(SymbolToken.LParen);
        const expressions = this.parseExpressions();
        this.scanRequired(SymbolToken.RParen);
        return new AstParenthesizedExpression(expressions);
      case RegExpToken.Identifier:
        if (this.scanner.peek(1).kind === SymbolToken.LParen) {
          const identifier = this.scanner.read();
          this.scanRequired(SymbolToken.LParen);
          const invocation = new AstInvocationExpression(
            new AstLiteralIdentifier(identifier),
            this.parseExpressions()
          );
          this.scanRequired(SymbolToken.RParen);
          return invocation;
        }
        return new AstLiteralIdentifier(this.scanner.read());
      case RegExpToken.LiteralBoolean:
        return new AstLiteralBoolean(this.scanner.read());
      case RegExpToken.LiteralNumber:
        return new AstLiteralNumber(this.scanner.read());
      case RegExpToken.LiteralString:
        return new AstLiteralString(this.scanner.read());
      default:
        return undefined;
    }
  }

  private parseFunctionDeclaration(): AstFunctionDeclaration {
    const identifier = this.scanner.lastMatch[0];
    this.scanRequired(StringToken.Arrow);
    if (this.scanOptional(SymbolToken.LCurly)) {
      const fn = new AstFunctionDeclaration(
        identifier,
        this.parseExpressions()
      );
      this.scanRequired(SymbolToken.RCurly);
      return fn;
    }
    const expression = this.parseExpression()!;
    return new AstFunctionDeclaration(identifier, [expression]);
  }

  private parseExpressions(): AstExpression[] {
    const expressions: AstExpression[] = [];
    let expression = this.parseExpression();
    while (expression) {
      expressions.push(expression);
      expression = this.parseExpression();
    }
    return expressions;
  }

  private scanOptional(token: TokenKind): Token | undefined {
    if (this.scanner.scan(token)) {
      return this.scanner.lastMatch[0];
    }
    return undefined;
  }

  private scanRequired(token: TokenKind): void {
    /* istanbul ignore next */
    if (!this.scanner.scan(token)) {
      // The lexer should prevent this happening in a typical program.
      /* istanbul ignore next */
      const next = this.scanner.peek().kind;
      /* istanbul ignore next */
      throw new SyntaxError(`Expected ${token.name} got ${next.name}.`);
    }
  }
}
