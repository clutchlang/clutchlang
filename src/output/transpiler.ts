import {
  AstCompilationUnit,
  AstFunctionDeclaration,
  AstInvocationExpression,
  AstLiteralBoolean,
  AstLiteralIdentifier,
  AstLiteralNumber,
  AstLiteralString,
  AstParenthesizedExpression,
  AstVariableDeclaration,
} from '../parser/source/ast/node';
import { AstVisitor } from '../parser/source/ast/visitor';

/**
 * A basic "compiler" that does a literal translation to JavaScript.
 *
 * It does *not*:
 * - Implement the Clutch language specification.
 * - Perform validation, type-checking, or any other error checking.
 *
 * This is a stop-gap output type until the pipeline is more complete.
 */
export class JsOutputTranspiler extends AstVisitor {
  public visitCompilationUnit(node: AstCompilationUnit): string {
    return node.functions.map(e => e.visit(this)).join('\n');
  }

  public visitFunctionDeclaration(node: AstFunctionDeclaration): string {
    return `
      function ${node.name}(${node.parameters.map(e => e.name).join(', ')}) {
        ${node.body.map(e => `${e.visit(this)};`).join('\n')}
      }
      ${node.name === 'main' ? 'main();' : ''}
    `;
  }

  public visitVariableDeclaration(node: AstVariableDeclaration): string {
    return `let ${node.name} = ${node.value.visit(this)}`;
  }

  public visitLiteralBoolean(node: AstLiteralBoolean): string {
    return node.value.toString();
  }

  public visitLiteralNumber(node: AstLiteralNumber): string {
    return node.value.toString();
  }

  public visitLiteralString(node: AstLiteralString): string {
    return `'${node.value.toString()}'`;
  }

  public visitLiteralIdentifier(node: AstLiteralIdentifier): string {
    return node.name.toString();
  }

  public visitInvocationExpression(node: AstInvocationExpression): string {
    const target = node.target;
    // Special cased for now.
    if (target instanceof AstLiteralIdentifier && target.name === 'print') {
      return `console.log(${node.args[0].visit(this)})`;
    }
    return `${target.visit(this)}(${node.args
      .map(e => e.visit(this))
      .join(', ')})`;
  }

  public visitParenthesizedExpression(
    node: AstParenthesizedExpression
  ): string {
    return `(${node.body.map(e => e.visit(this)).join(', ')})`;
  }
}
