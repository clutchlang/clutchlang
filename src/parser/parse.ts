import { AstCompilationUnit } from './source/ast/node';
import { AstParser } from './source/ast/parser';
import { TokenScanner } from './source/ast/scanner';
import { Lexer } from './source/tokenizer/lexer';
import { SourceScanner } from './source/tokenizer/scanner';

export function parse(program: string): AstCompilationUnit {
  const scanner = new SourceScanner(program);
  const lexer = new Lexer(scanner);
  const tokens = Array.from(lexer);
  const parser = new AstParser(new TokenScanner(tokens));
  return parser.parseCompilationUnit();
}
