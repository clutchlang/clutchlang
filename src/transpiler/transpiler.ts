// tslint:disable:no-magic-numbers

import { StringBuffer } from '../agnostic/strings';
import * as ast from '../language/parser';

// TODO: Refactor out a RecursiveAstVisitor, GeneralizingAstVisitor.

/**
 * Implements a simple/non-compliant Clutch-to-JS transpiler.
 *
 * It does not perform any semantic analysis, and is not able to support the
 * entire language, and instead is intended as an (early) demo and a way to test
 * the parser.
 */
export class SimpleJsTranspiler extends ast.AstVisitor<
  StringBuffer,
  StringBuffer
> {
  public visitFileRoot(
    node: ast.FileRoot,
    context = new StringBuffer()
  ): StringBuffer {
    node.topLevelElements.forEach(e => e.accept(this, context));
    return context;
  }

  public visitBinaryExpression(
    node: ast.BinaryExpression,
    context = new StringBuffer()
  ): StringBuffer {
    node.left.accept(this, context);
    context.write(` ${node.operatorToken.lexeme} `);
    node.right.accept(this, context);
    return context;
  }

  public visitGroupExpression(
    node: ast.GroupExpression,
    context = new StringBuffer()
  ): StringBuffer {
    context.write('(');
    node.expression.accept(this, context);
    context.write(')');
    return context;
  }

  public visitConditionalExpression(
    node: ast.ConditionalExpression,
    context = new StringBuffer()
  ): StringBuffer {
    // TODO: Support statement blocks?
    node.condition.accept(this, context);
    context.write(' ? ');
    (node.body as ast.Expression).accept(this, context);
    if (node.elseToken) {
      context.write(' : ');
      (node.elseBody as ast.Expression).accept(this, context);
    }
    return context;
  }

  public visitInvokeExpression(
    node: ast.InvokeExpression,
    context = new StringBuffer()
  ): StringBuffer {
    node.target.accept(this, context);
    context.write('(');
    node.parameters.forEach((e, i) => {
      e.accept(this, context);
      if (i < node.parameters.length - 1) {
        context.write(', ');
      }
    });
    context.write(')');
    return context;
  }

  public visitLiteralBoolean(
    node: ast.LiteralBoolean,
    context = new StringBuffer()
  ): StringBuffer {
    context.write(node.value);
    return context;
  }

  public visitLiteralNumber(
    node: ast.LiteralNumber,
    context = new StringBuffer()
  ): StringBuffer {
    context.write(node.value);
    return context;
  }

  public visitLiteralString(
    node: ast.LiteralString,
    context = new StringBuffer()
  ): StringBuffer {
    context.write("'");
    context.write(node.value);
    context.write("'");
    return context;
  }

  public visitSimpleName(
    node: ast.LiteralIdentifier,
    context = new StringBuffer()
  ): StringBuffer {
    context.write(node.name);
    return context;
  }

  public visitUnaryExpression(
    node: ast.UnaryExpression,
    context = new StringBuffer()
  ): StringBuffer {
    if (node.isPrefix) {
      context.write(node.operatorToken.lexeme);
      node.target.accept(this, context);
    } else {
      node.target.accept(this, context);
      context.write(node.operatorToken.lexeme);
    }
    return context;
  }

  public visitReturnStatement(
    node: ast.ReturnStatement,
    context = new StringBuffer()
  ): StringBuffer {
    context.write('return ');
    if (node.expression) {
      node.expression.accept(this, context);
    }
    return context;
  }

  public visitVariableDeclarationStatement(
    node: ast.VariableDeclarationStatement,
    context = new StringBuffer()
  ): StringBuffer {
    context.write('let ');
    node.name.accept(this, context);
    context.write(' = ');
    node.expression.accept(this, context);
    return context;
  }

  public visitFunctionDeclaration(
    node: ast.FunctionDeclaration,
    context = new StringBuffer()
  ): StringBuffer {
    context.write('function ');
    node.name.accept(this, context);
    context.write('(');
    node.parameters.forEach((e, i, a) => {
      const last = i === a.length - 1;
      e.accept(this, context);
      if (!last) {
        context.write(', ');
      }
    });
    context.write(') ');
    context.indent(2);
    context.writeLine('{');
    if (node.body instanceof ast.StatementBlock) {
      node.body.statements.forEach((e, i, a) => {
        const last = i === a.length - 1;
        if (last && e instanceof ast.Expression) {
          context.writeIndented('return ');
        } else {
          context.writeIndented();
        }
        e.accept(this, context);
        context.write(';\n');
      });
    } else {
      context.writeIndented('return ');
      node.body.accept(this, context);
      context.write(';\n');
    }
    context.indent(-2);
    context.writeLine('}');
    return context;
  }
}
