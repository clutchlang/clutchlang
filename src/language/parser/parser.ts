import { FileParser } from './parse/file';

export class ClutchParser extends FileParser {}
