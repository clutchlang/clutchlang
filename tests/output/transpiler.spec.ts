import { JsOutputTranspiler } from '../../src/output/transpiler';
import { parse } from '../../src/parser/parse';

describe('JsOutputTranspiler', () => {
  it('should support "Hello World"', () => {
    const program = parse(`
      main => {
        print('Hello World')
      }
    `);
    const visitor = new JsOutputTranspiler();
    expect(program.visit(visitor)).toMatchSnapshot();
  });

  it('should support most ast nodes', () => {
    const program = parse(`
      example => {
        true
        1
        'Hello'
        foobar
        method(false)
        f(1 2)
      }

      explicit => {
        return true
      }

      implicit => {
        true
      }

      noReturn => {
        let variable = 5
      }

      withIf => if true 1

      withIfElse => if true 1 else 2

      withIfStatement => {
        if (true) {
          print(1)
        } else {
          print(2)
        }
      }

      ternary(a b c) => if a b else c
    `);
    const visitor = new JsOutputTranspiler();
    expect(program.visit(visitor)).toMatchSnapshot();
  });

  it('should support parenthesized ast nodes', () => {
    const program = parse(`
      example => {
        (true false)
        (false true)
      }
    `);
    const visitor = new JsOutputTranspiler();
    expect(program.visit(visitor)).toMatchSnapshot();
  });

  it('should support functions with parameters', () => {
    const program = parse(`
      print2(a b) => {
        print(a)
        print(b)
      }
    `);
    const visitor = new JsOutputTranspiler();
    expect(program.visit(visitor)).toMatchSnapshot();
  });
});
