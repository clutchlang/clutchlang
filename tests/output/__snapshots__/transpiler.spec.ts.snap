// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JsOutputTranspiler should support "Hello World" 1`] = `
"function main() {
  return console.log(\\"Hello World\\");
}
main();
"
`;

exports[`JsOutputTranspiler should support functions with parameters 1`] = `
"function print2(a, b) {
  console.log(a);
  return console.log(b);
}
"
`;

exports[`JsOutputTranspiler should support most ast nodes 1`] = `
"function example() {
  true;
  1;
  (\\"Hello\\");
  foobar;
  method(false);
  return f(1, 2);
}

function explicit() {
  return true;
}

function implicit() {
  return true;
}

function noReturn() {
  let variable = 5;
}

function withIf() {
  return (function() {
    if (true) {
      return 1;
    }
  })();
}

function withIfElse() {
  return (function() {
    if (true) {
      return 1;
    } else {
      return 2;
    }
  })();
}

function withIfStatement() {
  return (function() {
    if (true) {
      return console.log(1);
    } else {
      return console.log(2);
    }
  })();
}

function ternary(a, b, c) {
  return (function() {
    if (a) {
      return b;
    } else {
      return c;
    }
  })();
}
"
`;

exports[`JsOutputTranspiler should support parenthesized ast nodes 1`] = `
"function example() {
  true, false;
  return false, true;
}
"
`;
